### PR TITLE
[SYCL] Float2Half precision requirements, convert method correction.

### PR DESCRIPTION
**Pull request by garimagu**
- Added the support for output operator for half datatypes.
- Edit test case vec_convert.cpp for half datatype.
  Corrected source code for float2Half conversion.
- Correct the automatic/rte mode implementation.

Signed-off-by: Garima Gupta <garima.gupta@intel.com>

### CODE REVIEW

File: sycl/source/half_type.cpp [hunk 1]
```diff
@@ -9,8 +9,8 @@
 #include <CL/sycl/half_type.hpp>
 // This is included to enable __builtin_expect()
 #include <CL/sycl/detail/platform_util.hpp>
-#include <iostream>
 #include <cstring>
+#include <iostream>
```
Comment (alexeyvoronov) [inline sycl/source/half_type.cpp]: Why do you need this include?

File: sycl/include/CL/sycl/types.hpp [hunk 2]
```diff
@@ -273,13 +269,22 @@ template <typename T, typename R, rounding_mode roundingMode>
 detail::enable_if_t<!std::is_same<T, R>::value && is_float_to_int<T, R>::value,
                     R>
 convertImpl(T Value) {
+	R Result;
```
Comment (alexeyvoronov) [inline sycl/include/CL/sycl/types.hpp]: These variables are used only in one case. Let's define them there.

File: sycl/include/CL/sycl/types.hpp [hunk 3]
```diff
@@ -258,6 +253,7 @@ detail::enable_if_t<std::is_same<T, R>::value, R> convertImpl(T Value) {
   return Value;
 }
 
+// Note for float to half conversions, static_cast calls the conversion operator implemented for host that takes care of the precision requirements.
```
Comment (alexeyvoronov) [inline sycl/include/CL/sycl/types.hpp]: This line is more than 80 symbols. Please fix.

File: sycl/include/CL/sycl/types.hpp (hunk 2 above)
Comment (garimagu) [reply sycl/include/CL/sycl/types.hpp]: If we do that, we get a compilation error saying we cannot bypass the definition of this variable, or some error on those lines.

File: sycl/source/half_type.cpp (hunk 1 above)
Comment (garimagu) [reply sycl/source/half_type.cpp]: We don't need it anymore after the removal of the iostream operators.

File: sycl/include/CL/sycl/types.hpp (hunk 2 above)
Comment (alexeyvoronov) [reply sycl/include/CL/sycl/types.hpp]: I think you will help brackets : https://godbolt.org/z/80_anA

Comment (garimagu) [reply sycl/include/CL/sycl/types.hpp]: Thanks. The godbolt link really helps.

File: sycl/include/CL/sycl/types.hpp [hunk 4]
```diff
@@ -50,17 +50,12 @@
 #include <CL/sycl/multi_ptr.hpp>
 
 #include <array>
+#include <cfenv>
```
Comment (bader) [inline sycl/include/CL/sycl/types.hpp]: Please, put this include under `#ifndef __SYCL_DEVICE_ONLY__`.

File: sycl/include/CL/sycl/types.hpp [hunk 5]
```diff
@@ -278,8 +275,17 @@ convertImpl(T Value) {
     // Round to nearest even is default rounding mode for floating-point types
   case rounding_mode::automatic:
     // Round to nearest even.
-  case rounding_mode::rte:
-    return std::round(Value);
+  case rounding_mode::rte: {
+    int OldRoundingDirection = std::fegetround();
+    int Err = std::fesetround(FE_TONEAREST);
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/types.hpp]: C++ reference says we have to set pragma FENV_ACCESS to safely use this.
http://www.cplusplus.com/reference/cfenv/fesetround/

File: sycl/include/CL/sycl/types.hpp (hunk 4 above)
Comment (garimagu) [reply sycl/include/CL/sycl/types.hpp]: done.

File: sycl/include/CL/sycl/types.hpp (hunk 5 above)
Comment (garimagu) [reply sycl/include/CL/sycl/types.hpp]: Thank you for noticing this. 
I added this statement.

**Comment by garimagu** [comment]
> ```c++
> // float to int
> template <typename T, typename R, rounding_mode roundingMode>
> detail::enable_if_t<!std::is_same<T, R>::value && is_float_to_int<T, R>::value,
>                     R>
> ```
> 
> Do we need `!std::is_same<T, R>::value` check? Isn't it always `true` if `is_float_to_int<T, R>::value == true`?

You are right. I think it was used to follow a certain flow of the code.
Removed it.

**Comment by garimagu** [comment]
@bader Please see if you feel anything else is needed.

**Comment by bader** [comment]
> @bader Please see if you feel anything else is needed.

This version looks good. Thanks!

**Comment by garimagu** [comment]
The failing tests are unexpected passes on Windows:
boolean.cpp and vector_relational.cpp.
Is this a known issue?

**Comment by alexeyvoronov** [comment]
> The failing tests are unexpected passes on Windows:
> boolean.cpp and vector_relational.cpp.
> Is this a known issue?
Yes. You just needed to re-base your patch.

**Pull request merged**